### DISCUSSION TITLE
virtualenv/venv selector in jupyterlab

### DISCUSSION THREAD

**Discussion by bluss** [jupyterlab/jupyterlab#16641] [General]
Hi Everyone,

I wanted to ask for previous discussions or clarifications on virtualenv selection - why do we have the current state of features related to Python virtualenvs?

- VSCodium and VS Code have a virtualenv selector for the **ipykernel** environment
- Why doesn't JupyterLab have this selector for the **ipykernel** environment?

I know the Python community is quite diverse in how it's working and using python environments (in the general sense, not just venvs). I barely know how to work with conda, but I know that many use that and so on. The Python package landscape is changing quickly from year to year too (different discussion..).

I know there already exists “competing” solutions

- Installing kernelspecs, one per virtualenv (or other env)
- People use virtualenvs (or other env) for the whole combination of JupyterLab + Kernel
- JupyterLab Desktop has a nice virtualenv selector interface (again from the whole combination of JupyterLab + Kernel)

I've been working on [pyproject-local-kernel](https://bluss.github.io/pyproject-local-kernel/) which is a “new”[^1] way to configure the python environments for notebooks in jupyterlab - without having to install so many kernelspecs.

I'm not here to push for my solution - if jupyterlab just included a kernel virtualenv selector then my solution would be redundant. But I wanted to ask the state of things and accepted knowledge - and how can we improve it from this point.

I use my solution at work. The reason is reproducibility and ease of setup - can others check out the code from git and just start working on the notebooks? If we skip custom setup steps like installing kernelspecs or setting up bespoke environments, this gets easier. /A virtualenv selector solution will be sucessful if it has the same properites./

[^1]: [Well, someone else did it first](https://github.com/pathbird/poetry-kernel)

**Comment by krassowski** [comment]
> - VSCodium and VS Code have a virtualenv selector for the ipykernel environment
> - Why doesn't JupyterLab have this selector for the ipykernel environment?

Well, you already know that [JupyterLab Desktop](https://github.com/jupyterlab/jupyterlab-desktop) does have an environment selector. Looking at it otherwise is not comparing apples to apples. Plain JupyterLab is used in different setups and in many there may be no virtual environment concept at all. This is especially the case for usage outside of the Python world.

**Reply by bluss** [reply]
The JupyterLab Desktop virtualenv selector is not what I would consider the same - it's for all of jupyterlab, not just the kernel. It's possible for me to have two different notebooks open in the same jupyterlab instance, and want to have different dependencies in each of them.

**Reply by bluss** [reply]
I'm not here to complain or drag the project down, quite the contrary. I'm trying to gather some information before filing an issue.

I think that you have heard the feature request "please give us a virtualenv selector" before and I'm trying to understand what the official jupyterlab stance is on that, so that I don't waste time on retracing old discussions. However, I can't find any old discussions about it on github, do you know if it has been discussed?

**Reply by krassowski** [reply]
I would say go ahead and file an issue, I don't recall a recent one in JupyterLab (I might be wrong). Discussions do not receive a lot of attention, and are not triaged.

> so that I don't waste time on retracing old discussions

Old discussions are old, things may have changed :)

One recent-ish issue that I know of is https://github.com/jupyterlab/jupyterlab/issues/12916 but its not about the selector itself. I would love to have a selector as an extension, possibly developed independently and graduated into JupyterLab org. Maybe we would want to include it by default maybe not. To get a good indication of the sentiment opening an issue or attending a weekly call on Wednesday may be a good idea https://github.com/jupyterlab/frontends-team-compass?tab=readme-ov-file#weekly-jupyter-frontends-team-meeting

But in general there may be past discussions in https://github.com/jupyter/notebook or https://github.com/ipython/ipython repo. In general major changes on protocol level go through JEP process https://github.com/jupyter/enhancement-proposals

**Reply by krassowski** [reply]
> In general major changes on protocol level go through JEP process

Having an extension would not be a major change; having it built in might require a major change.

**Reply by bluss** [reply]
Thanks for the detailed answer! I see that you talk about environments, and my search for "virtualenv" gave nothing, and it's probably some simple explanation like that, different terms used for those (similar but not the same) concepts.

**Reply by krassowski** [reply]
One interesting question is: can we have a generic UI selector that would work for python venvs (whatever we call them) and say R environments `packrat`/`renv` or Julia versions/envs.
